### PR TITLE
nixVersions.nix_2_20: init at 2.20.5; nixVersions.unstable: 2.19 -> 2.20

### PR DESCRIPTION
**Pull request by SuperSandro2000**
## Description of changes

Fails to compile https://github.com/NixOS/nix/issues/9888

```
nix> In file included from include/boost/container/vector.hpp:27,
nix>                  from include/boost/container/small_vector.hpp:27,
nix>                  from src/libexpr/gc-small-vector.hh:3,
nix>                  from src/libexpr/eval.cc:21:
nix> include/boost/container/allocator_traits.hpp: In instantiation of 'struct boost::container::allocator_traits<traceable_allocator<void> >':
nix> include/boost/container/small_vector.hpp:122:7:   required from 'class boost::container::small_vector_allocator<nix::Value*, traceable_allocator<void>, void>'
nix> include/boost/container/allocator_traits.hpp:145:48:   required from 'struct boost::container::allocator_traits<boost::container::small_vector_allocator<nix::Value*, traceable_allocator<void>, void> >'
nix> include/boost/container/vector.hpp:802:88:   required from 'class boost::container::vector<nix::Value*, boost::container::small_vector_allocator<nix::Value*, traceable_allocator<void>,
void>, void>'
nix> include/boost/container/small_vector.hpp:341:7:   required from 'class boost::container::small_vector_base<nix::Value*, traceable_allocator<nix::Value*>, void>'
nix> include/boost/container/small_vector.hpp:531:7:   required from 'class boost::container::small_vector<nix::Value*, 4, traceable_allocator<nix::Value*>, void>'
nix> src/libexpr/eval.cc:1730:31:   required from here
nix> include/boost/container/allocator_traits.hpp:145:48: error: 'typedef void traceable_allocator<void>::value_type' is private within this context
nix> In file included from src/libexpr/value.hh:15,
nix>                  from src/libexpr/nixexpr.hh:7,
nix>                  from src/libexpr/attr-set.hh:4,
nix>                  from src/libexpr/eval.hh:4,
nix>                  from src/libexpr/eval.cc:1:
nix> /nix/store/7ikfgkswxfcqcxhpcz3bcpg98hv523y4-boehm-gc-8.2.4-dev/include/gc/gc_allocator.h:319:23: note: declared private here
nix>   319 |   typedef void        value_type;
nix>       |                       ^~~~~~~~~~
nix> include/boost/container/allocator_traits.hpp:210:13: error: 'typedef void traceable_allocator<void>::value_type' is private within this context
nix> /nix/store/7ikfgkswxfcqcxhpcz3bcpg98hv523y4-boehm-gc-8.2.4-dev/include/gc/gc_allocator.h:319:23: note: declared private here
nix>   319 |   typedef void        value_type;
nix>       |                       ^~~~~~~~~~
nix> include/boost/container/allocator_traits.hpp:215:16: error: 'typedef void traceable_allocator<void>::value_type' is private within this context
nix> /nix/store/7ikfgkswxfcqcxhpcz3bcpg98hv523y4-boehm-gc-8.2.4-dev/include/gc/gc_allocator.h:319:23: note: declared private here
nix>   319 |   typedef void        value_type;
nix>       |                       ^~~~~~~~~~
nix> include/boost/container/allocator_traits.hpp:215:16: error: 'typedef void traceable_allocator<void>::value_type' is private within this context
nix> /nix/store/7ikfgkswxfcqcxhpcz3bcpg98hv523y4-boehm-gc-8.2.4-dev/include/gc/gc_allocator.h:319:23: note: declared private here
nix>   319 |   typedef void        value_type;
nix>       |                       ^~~~~~~~~~
nix> include/boost/container/allocator_traits.hpp:219:13: error: 'typedef void traceable_allocator<void>::value_type' is private within this context
nix> /nix/store/7ikfgkswxfcqcxhpcz3bcpg98hv523y4-boehm-gc-8.2.4-dev/include/gc/gc_allocator.h:319:23: note: declared private here
nix>   319 |   typedef void        value_type;
nix>       |                       ^~~~~~~~~~
nix> include/boost/container/allocator_traits.hpp:223:16: error: 'typedef void traceable_allocator<void>::value_type' is private within this context
nix> /nix/store/7ikfgkswxfcqcxhpcz3bcpg98hv523y4-boehm-gc-8.2.4-dev/include/gc/gc_allocator.h:319:23: note: declared private here
nix>   319 |   typedef void        value_type;
nix>       |                       ^~~~~~~~~~
nix> include/boost/container/allocator_traits.hpp:228:16: error: 'typedef void traceable_allocator<void>::value_type' is private within this context
nix> /nix/store/7ikfgkswxfcqcxhpcz3bcpg98hv523y4-boehm-gc-8.2.4-dev/include/gc/gc_allocator.h:319:23: note: declared private here
nix>   319 |   typedef void        value_type;
nix>       |                       ^~~~~~~~~~
nix> include/boost/container/allocator_traits.hpp:233:16: error: 'typedef void traceable_allocator<void>::value_type' is private within this context
nix> /nix/store/7ikfgkswxfcqcxhpcz3bcpg98hv523y4-boehm-gc-8.2.4-dev/include/gc/gc_allocator.h:319:23: note: declared private here
nix>   319 |   typedef void        value_type;
nix>       |                       ^~~~~~~~~~
nix> In file included from include/boost/container/vector.hpp:27,
nix>                  from include/boost/container/small_vector.hpp:27,
nix>                  from src/libexpr/gc-small-vector.hh:3,
nix>                  from src/libexpr/primops.cc:7:
nix> include/boost/container/allocator_traits.hpp: In instantiation of 'struct boost::container::allocator_traits<traceable_allocator<void> >':
nix> include/boost/container/small_vector.hpp:122:7:   required from 'class boost::container::small_vector_allocator<nix::Value*, traceable_allocator<void>, void>'
nix> include/boost/container/allocator_traits.hpp:145:48:   required from 'struct boost::container::allocator_traits<boost::container::small_vector_allocator<nix::Value*, traceable_allocator<void>, void> >'
nix> include/boost/container/vector.hpp:802:88:   required from 'class boost::container::vector<nix::Value*, boost::container::small_vector_allocator<nix::Value*, traceable_allocator<void>,
void>, void>'
nix> include/boost/container/small_vector.hpp:341:7:   required from 'class boost::container::small_vector_base<nix::Value*, traceable_allocator<nix::Value*>, void>'
nix> include/boost/container/small_vector.hpp:531:7:   required from 'class boost::container::small_vector<nix::Value*, 128, traceable_allocator<nix::Value*>, void>'
nix> src/libexpr/primops.cc:2733:56:   required from here
nix> include/boost/container/allocator_traits.hpp:145:48: error: 'typedef void traceable_allocator<void>::value_type' is private within this context
nix> In file included from src/libexpr/value.hh:15,
nix>                  from src/libexpr/nixexpr.hh:7,
nix>                  from src/libexpr/attr-set.hh:4,
nix>                  from src/libexpr/eval.hh:4,
nix>                  from src/libexpr/eval-inline.hh:5,
nix>                  from src/libexpr/primops.cc:4:
nix> /nix/store/7ikfgkswxfcqcxhpcz3bcpg98hv523y4-boehm-gc-8.2.4-dev/include/gc/gc_allocator.h:319:23: note: declared private here
nix>   319 |   typedef void        value_type;
nix>       |                       ^~~~~~~~~~
nix> include/boost/container/allocator_traits.hpp:210:13: error: 'typedef void traceable_allocator<void>::value_type' is private within this context
nix> /nix/store/7ikfgkswxfcqcxhpcz3bcpg98hv523y4-boehm-gc-8.2.4-dev/include/gc/gc_allocator.h:319:23: note: declared private here
nix>   319 |   typedef void        value_type;
nix>       |                       ^~~~~~~~~~
nix> include/boost/container/allocator_traits.hpp:215:16: error: 'typedef void traceable_allocator<void>::value_type' is private within this context
nix> /nix/store/7ikfgkswxfcqcxhpcz3bcpg98hv523y4-boehm-gc-8.2.4-dev/include/gc/gc_allocator.h:319:23: note: declared private here
nix>   319 |   typedef void        value_type;
nix>       |                       ^~~~~~~~~~
nix> include/boost/container/allocator_traits.hpp:215:16: error: 'typedef void traceable_allocator<void>::value_type' is private within this context
nix> /nix/store/7ikfgkswxfcqcxhpcz3bcpg98hv523y4-boehm-gc-8.2.4-dev/include/gc/gc_allocator.h:319:23: note: declared private here
nix>   319 |   typedef void        value_type;
nix>       |                       ^~~~~~~~~~
nix> include/boost/container/allocator_traits.hpp:219:13: error: 'typedef void traceable_allocator<void>::value_type' is private within this context
nix> /nix/store/7ikfgkswxfcqcxhpcz3bcpg98hv523y4-boehm-gc-8.2.4-dev/include/gc/gc_allocator.h:319:23: note: declared private here
nix>   319 |   typedef void        value_type;
nix>       |                       ^~~~~~~~~~
nix> include/boost/container/allocator_traits.hpp:223:16: error: 'typedef void traceable_allocator<void>::value_type' is private within this context
nix> /nix/store/7ikfgkswxfcqcxhpcz3bcpg98hv523y4-boehm-gc-8.2.4-dev/include/gc/gc_allocator.h:319:23: note: declared private here
nix>   319 |   typedef void        value_type;
nix>       |                       ^~~~~~~~~~
nix> include/boost/container/allocator_traits.hpp:228:16: error: 'typedef void traceable_allocator<void>::value_type' is private within this context
nix> /nix/store/7ikfgkswxfcqcxhpcz3bcpg98hv523y4-boehm-gc-8.2.4-dev/include/gc/gc_allocator.h:319:23: note: declared private here
nix>   319 |   typedef void        value_type;
nix>       |                       ^~~~~~~~~~
nix> include/boost/container/allocator_traits.hpp:233:16: error: 'typedef void traceable_allocator<void>::value_type' is private within this context
nix> /nix/store/7ikfgkswxfcqcxhpcz3bcpg98hv523y4-boehm-gc-8.2.4-dev/include/gc/gc_allocator.h:319:23: note: declared private here
nix>   319 |   typedef void        value_type;
nix>       |                       ^~~~~~~~~~
nix> make: *** [mk/patterns.mk:2: src/libexpr/eval.o] Error 1
nix> make: *** Waiting for unfinished jobs....
nix> make: *** [mk/patterns.mk:2: src/libexpr/primops.o] Error 1
nix> rm src/nix/generated-doc/files/profiles.md
```

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by infinisil** [comment]
Btw the `lib` test failure is legit, it detected a regression: https://github.com/NixOS/nix/issues/9901

**Comment by nyabinary** [comment]
What work does this need to be ready?

**Comment by infinisil** [comment]
@nyabinary Nixpkgs CI runs the `lib` tests for all `nixVersions`, so the regression in 2.20 should be fixed first.

**Comment by SuperSandro2000** [comment]
we are waiting for https://github.com/NixOS/nix/pull/10023 to be included in a tag

**Comment by edolstra** [comment]
2.20.2 is now released.

**Review by roberth — APPROVED** [review]

**Review by teutat3s — APPROVED** [review]

**Comment by infinisil** [comment]
I'm both glad and sad to see that the `lib.fileset` tests discovered more regressions..

- https://github.com/NixOS/nix/issues/10039
- https://github.com/NixOS/nix/issues/10040

**Comment by Ma27** [comment]
Personally, I'm not convinced that we should provide a Nix that's not compatible with our lib.

**Comment by SuperSandro2000** [comment]
Maybe nix should run those tests, so that in the future those regressions are not only noticed after a release and the bump in nixpkgs.

**Comment by infinisil** [comment]
@SuperSandro2000 Agreed, there's even already a draft PR by Robert for that: https://github.com/NixOS/nix/pull/9900

**Comment by roberth** [comment]
> nix should run those tests

We did, but our Nixpkgs pin lagged behind the fileset additions. Fix in https://github.com/NixOS/nix/pull/9900

-------- Original Message --------

File: pkgs/tools/package-management/nix/default.nix
Comment (alois31) [inline pkgs/tools/package-management/nix/default.nix:259]: What about the read-only regression fix (https://github.com/NixOS/nix/issues/10040)?

Comment (Lord-Valen) [reply pkgs/tools/package-management/nix/default.nix:259]: I believe we're waiting fo the next tag.

Comment (SuperSandro2000) [reply pkgs/tools/package-management/nix/default.nix:259]: we can fetch the patch and when I am not doing a typo, the tests should succeed.

**Comment by edolstra** [comment]
2.20.2 is now released, with fixes for the fetchGit issues.

File: pkgs/tools/package-management/nix/default.nix
Comment (kirillrdy) [reply pkgs/tools/package-management/nix/default.nix:264]: same here

**Comment by SuperSandro2000** [comment]
How did we encounter the next issue.... :notlikethis:

```
Running lib/tests/filesystem.sh
trace: warning: buildVimPluginFrom2Nix is deprecated: use buildVimPlugin instead
trace: warning: tryEval failed at: vimUtils.vimWithRC
tests ok
trace: warning: pkgs.writers.makeDataWriter is deprecated. Use pkgs.writeTextFile.
Running lib/tests/sources.sh
trace: warning: tryEval failed at: yabai
tests ok
Running lib/fileset/tests.sh
error:
       … from call site
         at «string»:11:3:
           10|   with lib.fileset;
           11|  (toSource { root = ./.; fileset = gitTrackedWith { } ./.; })
             |   ^

       … while calling 'toSource'
         at /build/lib/fileset/default.nix:338:14:
          337|   */
          338|   toSource = {
             |              ^
          339|     /*

       … from call site
         at /build/lib/fileset/default.nix:371:22:
          370|       filesetFilesystemRoot = (splitRoot fileset._internalBase).root;
          371|       sourceFilter = _toSourceFilter fileset;
             |                      ^
          372|     in

       … while calling '_toSourceFilter'
         at /build/lib/fileset/internal.nix:396:21:
          395|   # Type: fileset -> (String -> String -> Bool)
          396|   _toSourceFilter = fileset:
             |                     ^
          397|     let

       … from call site
         at /build/lib/fileset/internal.nix:400:14:
          399|       # which has the effect that they aren't included in the result
          400|       tree = _normaliseTreeFilter fileset._internalBase fileset._internalTree;
             |              ^
          401|

       … while calling '_normaliseTreeFilter'
         at /build/lib/fileset/internal.nix:272:32:
          271|   */
          272|   _normaliseTreeFilter = path: tree:
             |                                ^
          273|     if tree == "directory" || isAttrs tree then

       … from call site
         at /build/lib/fileset/internal.nix:867:8:
          866|     _create localPath
          867|       (recurse storePath);
             |        ^
          868|

       … while calling 'recurse'
         at /build/lib/fileset/internal.nix:858:17:
          857|     let
          858|       recurse = focusedStorePath:
             |                 ^
          859|         mapAttrs (name: type:

       error: access to path '/' is forbidden because it is not under Git control; maybe you should 'git add' it to the repository '/build/tmp.6dIU8Y46cx/work/shallow'?
test case at lib/fileset/tests.sh:1450 failed: toSource { root = ./.; fileset = gitTrackedWith { } ./.; } failed to evaluate, but it was expected to succeed
error: builder for '/nix/store/bgv903fhcl3zmc261hcvn1n2zsb3fi4p-nixpkgs-lib-tests-nix-2.20.3.drv' failed with exit code 1;
```

> nix should run those tests
> We did, but our Nixpkgs pin lagged behind the fileset additions. Fix in [NixOS/nix#9900](https://github.com/NixOS/nix/pull/9900)

Could this be added with priority, please? Otherwise 2.21 might be a similar odyssey

**Comment by infinisil** [comment]
I also reported this problem in https://github.com/NixOS/nix/issues/10039, though it looks like it was closed without getting fixed.. I did report two separate errors in the same issue since they're very similar

**Comment by alois31** [comment]
I think as stated above that the patch didn't make it to 2.20, so it needs to be restored.

**Comment by SuperSandro2000** [comment]
So what's the way forward here?

File: pkgs/tools/package-management/nix/default.nix
Comment (philiptaron) [inline pkgs/tools/package-management/nix/default.nix:266]: Hopefully this patch is short-lived.

**Review by philiptaron — APPROVED** [review]
As the GitHub mention tracker notes, I've been running this locally and it works great for me.

**Comment by RaitoBezarius** [comment]
Ideally, another patch release for the proposed patch here?

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
To avoid spurious merges.

**Review by roberth — CHANGES_REQUESTED** [review]
I've temporarily added 42c5c2b368c90ef559f3b459a6f0797ce2e48854 as an ad hoc pre-release check of sorts. Upstream's own checks will be improved, so this should be a one-off occurrence.
If CI passes here, we can make another patch release. Then rebase away that commit.

**Comment by SuperSandro2000** [comment]
> Ideally, another patch release for the proposed patch here?

That is out of my control. Can @edolstra do that?

**Comment by abathur** [comment]
https://github.com/NixOS/nix/issues/10116

**Comment by SuperSandro2000** [comment]
Not like we hadn't such issues before and needed to introduce workarounds, I don't consider this something blocking for this PR.

**Review by RaitoBezarius — DISMISSED** [review]
I don't like repeating myself: this has to wait for a **patch release** from Nix upstream. People wanting the ultimate bleeding edge can apply this PR as a patch themselves. There's no need to rush the `unstable` package.

For the stable package, this is out of the question given the current regression atmosphere.

**Comment by Ma27** [comment]
@RaitoBezarius For which patch specifically? edcb3430ef39a225aada06ef898c907d8277fbe8 and the stuff before that @roberth has added (and also appears to pass CI)?
That's part of 2.20.4: https://github.com/NixOS/nix/commits/2.20.4/
Sorry, wrote 2.20.5 first because I misread the commit log.

**Comment by RaitoBezarius** [comment]
> @RaitoBezarius For which patch specifically? edcb3430ef39a225aada06ef898c907d8277fbe8 and the stuff before that @roberth has added (and also appears to pass CI)? That's part of 2.20.4: [NixOS/nix@`2.20.4` (commits)](https://github.com/NixOS/nix/commits/2.20.4/) Sorry, wrote 2.20.5 first because I misread the commit log.

Yeah,
picture of the tag date

It seems recent, as long as this PR is bumped to this patch release and tests passes, we can proceed.

**Comment by SuperSandro2000** [comment]
> I assume you're strictly talking about the applied patch and not about the fact that `lib` tests were broken before?

I was referring to https://github.com/NixOS/nixpkgs/pull/285264#issuecomment-1971967804

> However, I'm against declaring it stable just in case people are considering to do so now.

Not part of this PR anyway and I can delegate that decision to other people.

> It seems recent, as long as this PR is bumped to this patch release and tests passes, we can proceed.

That somehow went passed me. Updated now.

**Comment by Ma27** [comment]
> Not part of this PR anyway and I can delegate that decision to other people.

Sure, that was statement was made proactively just in case people were about to reconsider this decision :)

**Review by roberth — APPROVED** [review]

File: pkgs/tools/package-management/nix/default.nix
Comment (Ma27) [inline pkgs/tools/package-management/nix/default.nix:260]: Should be `./patches/2_20`, right? DOesn't evaluate for me at least.

Comment (roberth) [reply pkgs/tools/package-management/nix/default.nix:260]: or update to 2.20.5

Comment (RaitoBezarius) [reply pkgs/tools/package-management/nix/default.nix:260]: yes... sorry

Comment (RaitoBezarius) [reply pkgs/tools/package-management/nix/default.nix:260]: feel free to update to 2.20.5 otherwise

Comment (Mic92) [reply pkgs/tools/package-management/nix/default.nix]: Dropped the patch by updating to 2.20.5

**Comment by Mic92** [comment]
I couldn't force push to squash the existing commits while fixing merge conflicts, so I created a new PR: https://github.com/NixOS/nixpkgs/pull/294120

**Comment by SuperSandro2000** [comment]
So, who wants to press the merge button?

**Comment by marcusramberg** [comment]
I guess the pr title should be updated to say 2.20.5.

**Comment by domenkozar** [comment]
Should this be backported to 23.11?

**Comment by RaitoBezarius** [comment]
Usually, we don't aim to backport `nixVersions.unstable` bump to stable.

Because unstable's Nix in stable _should be_ stable and too many people depend on unstable Nix thinking it's stable in general when it's not.

I have no strong feeling, but I'm not keeping the pieces if it breaks someone's production.

**Comment by Mic92** [comment]
What if the it's just a new attribute but leaving nixVersions.unstable untouched?

**Pull request merged**
